### DISCUSSION TITLE
Add :single_cassette option to RSpec metadata

### DISCUSSION THREAD

**Discussion by rachel-carvalho** [vcr/vcr#865] [Ideas]
### Background

I've been testing an API call with VCR that I can't really re-do (it works a bit like a delete) because I _have_ to create it manually and use the resulting `id` in my spec. It's something like this:

```ruby
describe '#delete', :vcr do
  let(:foo) { API::Foo.new('ID that I can only get manually') }

  subject { foo.delete }

  it 'changes status' do
    expect { subject }.to change foo, :status
  end

  it 'does something else' do
    # ...
  end

  it 'does a third thing' do
    # ...
  end
end
```

In this situation, each of the `it` blocks records its own cassette. But I need a new `id` for each new request, so when I'm re-recording my cassettes, the first `it` passes and the subsequent ones fail.

I could just copy/paste the first cassette with matching names for the other blocks, but it doesn't feel right.

Then I found I could use [the `:cassette_name` option](https://relishapp.com/vcr/vcr/docs/test-frameworks/usage-with-rspec-metadata) which ensures all my `it`s use the same cassette.

So I changed my specs:

```ruby
describe '#delete' do
  let(:foo) { API::Foo.new('ID that I can only get manually') }

  subject { foo.delete }

  context 'success', vcr: { cassette_name: 'API_Foo/_delete/success' } do
    it 'changes status' do
      expect { subject }.to change foo, :status
    end

    # ...
  end

  context 'failure', vcr: { cassette_name: 'API_Foo/_delete/failure' } do
    # ...
  end
end
```

But now I'm manually writing the cassette names to keep the same naming scheme as the automatically generated cassettes.

### Proposed solution

It would be really cool if we could specify we want a single cassette to be generated based on the block path, like this:

```ruby
describe '#delete' do
  subject { foo.delete }

  context 'success', vcr: { single_cassette: true } do
    # ...
  end

  context 'failure', vcr: { single_cassette: true } do
    # ...
  end
end
```

Sorry for the long wall of text and code, I couldn't find a succinct way to explain this.

I'd love have a go at this if you guys think it's a good idea (although I'll be leaving on vacation soon and will probably not have access to a computer for at least a month, but I'll do it then).

I realize there might be something I'm not doing right here so I'm open to suggestions that don't include adding a new feature :)

**Comment by mcmire** [comment]
I just came across into this and I'm curious if I want a similar feature than you. Would `single_cassette` generate a cassette for the entire context as a whole instead of each test individually?

**Comment by dieggofaustino** [comment]
I'm with the same problem right now and found this by searching for a nice solution. In my case all `it` inside a context should use the same cassette

**Comment by ixti** [comment]
This especially handy when you have bunch of `it { is_expected.to ... }` blocks.

**Comment by ixti** [comment]
I would like to note though, that proposed API seems a bit confusing although all my attempts to propose something better failed.

**Comment by mcmire** [comment]
I agree about the API. I'm wondering if `cassette_group` or `group_cassettes` would be a better name:

``` ruby
describe '#delete' do
  subject { foo.delete }

  context 'success', vcr: { cassette_group: true } do
    # ...
  end

  context 'failure', vcr: { group_cassettes: true } do
    # ...
  end
end
```

**Comment by hirowatari** [comment]
I absolutely love this feature and have been wanting it since I started using VCR at least 5 years ago.  I'd like to help push this over the finish line if that's is desired still (if not just let me know).

I believe what is preventing the merge is:
- the API name
- tests are failing
- documentation is missing

Is that true?

Regarding the API name, my opinion is that `group_cassettes: true` would be the best name.  `cassette_group` sounds like it could require a name like `cassette_name` does, whereas `group_cassettes` sounds like a question.  `single_cassette` is okay too, and `unique_cassette` might be slighter better, but `group_cassettes` seems best to me.

**Comment by mcmire** [comment]
@hirowatari A maintainer could weigh in better of course, but your assessment sounds right to me just from looking at this issue and the related PR.
